metadata/ota_metadata/metadata_gen.py

```diff
@@ -11,15 +13,47 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-
-#!/usr/bin/env python3
-
 import os
-from hashlib import sha256
 import argparse
 import pathlib
+import zstandard
 import igittigitt
+from hashlib import sha256
+
+ZSTD_COMPRESSION_ALG = "zstd"
+ZSTD_COMPRESSION_LEVEL = 10
+ZSTD_MULTITHREADS = 2
+CHUNK_SIZE = 4 * (1024**2)  # 4MiB
+
+
+def zstd_compress_file(
+    cctx: zstandard.ZstdCompressor,
+    src_fpath: str,
+    dst_fpath: str,
+    *,
+    cmpr_ratio: float,
+    filesize_threshold: int,
+) -> bool:
+    if (src_size := os.path.getsize(src_fpath)) < filesize_threshold:
+        return False  # skip file with too small size
+    try:
+        with open(src_fpath, "rb") as src_f, open(dst_fpath, "wb") as dst_f:
+            with cctx.stream_writer(dst_f, size=src_size) as compressor:
+                while data := src_f.read(CHUNK_SIZE):
+                    compressor.write(data)
+        # drop compressed file if cmpr ratio is too small or compressed failed
+        if (
+            not (compressed_bytes := os.path.getsize(dst_fpath))
+            or src_size / compressed_bytes < cmpr_ratio
+        ):
+            raise ValueError
```

Comment:
```
src_size / compressed_bytes < cmpr_ratio
```
In this case, raise ValueError seems strange, I think.

Reply:
Updated in [b2dc29a](https://github.com/tier4/ota-metadata/pull/12/commits/b2dc29a538de74ab3e55d21d4d6d1bbba8370c0e).

Reply:
No, I mean, the condition src_size / compressed_bytes < cmpr_ratio is normal, so raising exception seems strange for me.

Reply:
And we don't need to introduce new exception class. Any standard exception class is fine.

Reply:
Udated in [67c8fdd](https://github.com/tier4/ota-metadata/pull/12/commits/67c8fdd1737937dccc198824b82ccf8ec9a9898e).